ribbon-loadbalancer/src/main/java/com/netflix/loadbalancer/PollingServerListUpdater.java

```diff
@@ -0,0 +1,178 @@
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+
+/**
+ * A default strategy for the dynamic server list updater to update.
+ * (refactored and moved here from {@link com.netflix.loadbalancer.DynamicServerListLoadBalancer})
+ *
+ * @author David Liu
+ */
+public class PollingServerListUpdater implements ServerListUpdater {
+
+    private static final Logger logger = LoggerFactory.getLogger(PollingServerListUpdater.class);
+
+    private static long LISTOFSERVERS_CACHE_UPDATE_DELAY = 1000; // msecs;
+    private static int LISTOFSERVERS_CACHE_REPEAT_INTERVAL = 30 * 1000; // msecs;
+
+    private static class LazyHolder {
+        private final static String CORE_THREAD = "DynamicServerListLoadBalancer.ThreadPoolSize";
+        private final static DynamicIntProperty poolSizeProp = new DynamicIntProperty(CORE_THREAD, 2);
+        private static Thread _shutdownThread;
+
+        static ScheduledThreadPoolExecutor _serverListRefreshExecutor = null;
+
+        static {
+            int coreSize = poolSizeProp.get();
+            ThreadFactory factory = (new ThreadFactoryBuilder())
+                    .setNameFormat("PollingServerListUpdater-%d")
+                    .setDaemon(true)
+                    .build();
+            _serverListRefreshExecutor = new ScheduledThreadPoolExecutor(coreSize, factory);
+            poolSizeProp.addCallback(new Runnable() {
+                @Override
+                public void run() {
+                    _serverListRefreshExecutor.setCorePoolSize(poolSizeProp.get());
+                }
+
+            });
+            _shutdownThread = new Thread(new Runnable() {
+                public void run() {
+                    logger.info("Shutting down the Executor Pool for PollingServerListUpdater");
+                    shutdownExecutorPool();
+                }
+            });
+            Runtime.getRuntime().addShutdownHook(_shutdownThread);
+        }
+
+        private static void shutdownExecutorPool() {
+            if (_serverListRefreshExecutor != null) {
+                _serverListRefreshExecutor.shutdown();
+
+                if (_shutdownThread != null) {
+                    try {
+                        Runtime.getRuntime().removeShutdownHook(_shutdownThread);
+                    } catch (IllegalStateException ise) { // NOPMD
+                        // this can happen if we're in the middle of a real
+                        // shutdown,
+                        // and that's 'ok'
+                    }
+                }
+
+            }
+        }
+    }
+
+    private static ScheduledThreadPoolExecutor getRefreshExecutor() {
+        return LazyHolder._serverListRefreshExecutor;
+    }
+
+
+    private final AtomicLong lastUpdated = new AtomicLong(System.currentTimeMillis());
```

Comment:
Should be 'volatile' as we aren't doing any CAS or atomic increment operations here